### DISCUSSION TITLE
Injecting dependencies after the object has returned from the server

### DISCUSSION THREAD

**Discussion by GreatBarrier86** [MarimerLLC/csla#3804] [Questions]
Hi. Question. Right now, like the documentation shows, we’re defining the [Inject] attribute on the DP methods like usual. However, these business objects are sent to the server for save, and when the object is returned, obviously my dependencies are null. 

How am I supposed to “reinject” them on the returned object given that CSLA object constructors can’t have parameters?

**Comment by rockfordlhotka** [comment]
What dependencies are you hoping to access on the client?

**Comment by GreatBarrier86** [comment]
To name a few, the DataPortalFactory and some Azure utilities like a KeyVault encryption/decryption utility class. Plus any connection manager we want to create.

**Reply by rockfordlhotka** [reply]
Those type of services should only be accessed within a server side data portal operation.

**Reply by GreatBarrier86** [reply]
The data portal factory is necessary to actually call Fetch/Create...etc. That would then send the request to the server, but it has to be invoked from the client. 

As another example, let's say we had a SystemLog class that wrote to disk for whatever purpose. That needs to occur both client-side and server-side, as our management software reads those files and sends them to us. That's reasonable to have client-side. Or let's say that our connection manager class is also usable client-side to enable/disable client features based on what DB the server is connected to. In both of those cases, there is a legitimate use case for them being available client side.

I think the bigger point here is that in general, if the client has services that need to be accessed after save on the client-side, they'd need to be reinjected but there's no good entry point for doing so, unless i just missed it. The only thing i could imagine this scenario is to override DataPortalProxy.OnServerCompleteClient and recurse though the registered properties, applying the dependencies manually.

**Reply by StefanOssendorf** [reply]
There is a method iirc initialized or something like that which is called after the application context is set to a valid one. For some scenarios this might work.

I personally think supporting something like property injection would be great. Csla knows when a location change has happened and could populated those attributed properties/fields from the Di.

**Reply by GreatBarrier86** [reply]
@StefanOssendorf, so would you then suggest calling the IServiceProvider on the AppCtx in order to set the necessary dependencies? I suppose that would work but it doesn’t seem like the best solution. It would be better if the existing dependencies could just be reapplied after the BO is returned from the server.

**Reply by rockfordlhotka** [reply]
> The data portal factory is necessary to actually call Fetch/Create...etc

I don't understand this statement. Normally you inject an `IDataPortal<T>` service into your viewmodel, or page, or whatever. That's how you call the client-side data portal.

If you want, you can inject a data portal factory into your viewmodel, page, or whatever - which can be convenient if you need to make a lot of different root data portal calls. The thing is, in such a case you are probably _better_ served by using a _unit of work_ pattern and doing one root data portal call to a UoW object that can do all those multiple data portal calls while the code is executing on the server - that's almost always going to provide better performance.

**Reply by rockfordlhotka** [reply]
> SystemLog class that wrote to disk for whatever purpose

This might make some sense, though are you logging _inside_ your business classes? Like in rule types? Generally the only thing running in a business class on the client are properties and business rules.

Now if your desire is, on the client, to put logging into the client-side data portal pipeline, _that_ makes sense to me, and I've been thinking about creating a client-side interceptor model like we have on the server - specifically for logging purposes.

**Reply by rockfordlhotka** [reply]
> connection manager class is also usable client-side

Now this I _think_ I'm directly opposed to. The idea of logical client-side code interacting with a data store without using the data portal to get to logical server-side code is explicitly outside the architecture CSLA supports.

I understand the desire to have the data portal dynamically switch between local and remote modes based on whether the internet is available (for example), and you can do that with a custom data portal proxy that does exactly that. Having spent a lot of time in that space though, I personally feel like that's a poor architectural choice, and a better choice is to use the model used by outlook and every other email client out there: _always_ use a client-side data file, and have a separate agent/process that syncs the client-side data file with the server-side data file if the internet is available.

**Reply by rockfordlhotka** [reply]
@GreatBarrier86 I assume you've read the CSLA 6 migration doc?

https://github.com/MarimerLLC/csla/blob/main/docs/Upgrading%20to%20CSLA%206.md

Most notably the section on the client-side data portal and ways to use it now.

https://github.com/MarimerLLC/csla/blob/main/docs/Upgrading%20to%20CSLA%206.md#using-the-client-side-dataportal

**Reply by GreatBarrier86** [reply]
Rocky, yes I have read it. In our scenario, we have to support both a WinForms app with a DataPortalServer and multiple asp.net sites so basically we have to adopt DI given the multi-user nature and not just use statics. 

Regarding the services I’m referring to, I don’t dispute some of the designs are a little unique but we don’t have the luxury of going 100% appropriate DI. I recognize there may be superior ways in the long term, but right now, we’re trying to migrate to .NET 8 without upsetting the entire apple cart.

**Reply by rockfordlhotka** [reply]
Let me switch my mindset then. Rather than discussing how CSLA can be altered for a specific scenario, let's discuss how you can solve your scenario. A side-effect may be improvements to CSLA.

imo, the _ideal_ scenario would be to use DI in your WinForms codebase, allowing DI to create instances of the forms and their dependencies to be pretty much identical to aspnetcore or Blazor. That works _really well_, but is probably unrealistic in the context of an existing codebase.

Given that, I suspect your best bet is to create some services that you can access in WinForms via the service locator pattern (based on a static `App.ApplicationContext` property, and can inject in your aspnetcore code. These services may be similar to a viewmodel, or may be more like a set of domain type factories.

These factory types/services would be created via DI (again, using `ApplicationContext` or injection) and would abstract the use of the data portal. Within each factory you can inject and use `ILogger`, `IDataPortal`, etc.

I'm just brainstorming here, so might be off base, but that's my first thought from an abstraction perspective.

**Comment by GreatBarrier86** [comment]
Also, on the DataPortalFactory, we're doing it at a base level where we don't have the ability to work with a generic as a parameter, so we're using the DataPortalFactory. While the initial call may be from the ViewModel, it may be very possible that a child list is fetched on-demand AFTER the BO has been returned from Save or even the original fetch. Does that make sense?

1. Client ViewModel calls DataPortalFactory.GetPortal<Claim>().Fetch(myCrit)
2. Server fetches the object and injects the services in Fetch([Inject] Foo foo)
3. Server returns the object to the client
4. Client ViewModel receives the object.
5. Client needs to fetch a child list after user clicks Button1 to bind it to a DataGridView.

Given that we now need to do another fetch from within the BO, how would you expect us to do so without the dependency?

**Reply by rockfordlhotka** [reply]
You are using a lazy loading pattern to get the child on demand by using a command object to retrieve it? In that case, you'd use the `ApplicationContext.GetRequiredService` method to get the appropriate `IDataPortal<T>` for the command object.

**Comment by GreatBarrier86** [comment]
> You are using a lazy loading pattern to get the child on demand by using a command object to retrieve it?.

I have a BusinessBase object that has a child BusinessBindingListBase registered child object. When the BO is bound to the WinForms UI, after Fetch/Save, our DataGridView is bound to the child list object. So i guess you're saying that since each BO/BO List has a reference to ApplicationContext, any necessary dependencies should be requested through ApplicationContext.GetRequiredService?

**Comment by GreatBarrier86** [comment]
> imo, the _ideal_ scenario would be to use DI in your WinForms codebase, allowing DI to create instances of the forms and their dependencies to be pretty much identical to aspnetcore or Blazor. That works _really well_, but is probably unrealistic in the context of an existing codebase.

I've already rewritten all my form constructors to match your WinForms project example. It was better to just bite the bullet and do it. That part is working fine because I'm then able to use the injected DataPortalFactory to do GetPortal<T> from Form_Load to retrieve the BO. My issue is how to handle if the BO needs some of those injected dependencies AFTER it's been returned to the client. 

Possibly there is a concept I'm misunderstanding but imagine the following:
![image](https://github.com/MarimerLLC/csla/assets/29482393/34dc399c-a504-4b0e-8e0b-c9a94894232d)

**Comment by GreatBarrier86** [comment]
> Most notably the section on the client-side data portal and ways to use it now.
> https://github.com/MarimerLLC/csla/blob/main/docs/Upgrading%20to%20CSLA%206.md#using-the-client-side-dataportal

Here was my concern about this section. I'm fine following this for getting the IChildDataPortal specifically, but in a general sense, using a Service Locator is an anti-pattern: https://blog.ploeh.dk/2014/05/15/service-locator-violates-solid/#:~:text=The%20Service%20Locator%20anti-pattern%20violates%20the%20ISP%2C%20and%20thus,is%20that%20it%20violates%20encapsulation.

Given that, I was thinking that what should PROBABLY (?) happen is that upon return from the server, when you do the "if (IUsesApplicationContext)', since you're going through each object anyway, the dependencies should be reinjected then.

**Reply by rockfordlhotka** [reply]
Service locator is an anti-pattern, but it isn't clear to me that mobile objects can avoid it. The DI provider changes out from under BO objects, and any time that happens then all injected services become invalid.

So were we to do ctor or property injection on a BO type, we'd also have to figure out a way to "fix up" any service references you are holding any time the DI provider is swapped out.

This happens behind the `ApplicationContext` today, and so any time (in a method or property) you get a service via `ApplicationContext`, you are getting that service from the proper DI provider (and/or scope). And that's fine, because the reference to that service is per-method, and so we can be confident that the DI scope won't change out from under you as that method is running.

**Reply by rockfordlhotka** [reply]
Also, when lazy loading, you _must not_ directly use an `IChildDataPortal` - you _must_ use a command pattern so the fetch of the child object occurs on the logical server.

**Comment by GreatBarrier86** [comment]
> So were we to do ctor or property injection on a BO type, we'd also have to figure out a way to "fix up" any service references you are holding any time the DI provider is swapped out.

So maybe that's where I'm a little lost. Why can we not do ctor injection? I'm not entirely sure why that's off the table.

Regardless though, I would think a method similar in signature to fetch/create might do what we would need here. This obviously couldn't be an override, but it gives you the idea. Something that's called automatically during deserialization or even during Initialize, allowing the dependencies to be set as soon as possible.
`protected override void OnApplicationContextSet([Inject] IDataPortalFactory dataPortalFactory, [Inject] SystemLog loggingUtilities)`

**Reply by StefanOssendorf** [reply]
I would propose property injection. You could declare properties/fields within the BO to be set at the same time the application context is set. That would ensure you are always using the same scope as the application context.

**Reply by GreatBarrier86** [reply]
I’d like to know why ctor injection isn’t possible though. Obviously I’m missing something fundamental so I would appreciate the context.

**Reply by rockfordlhotka** [reply]
The core problem is that DI as a pattern or implementation was never designed around the idea of mobile objects. DI assumes everything is running in a memory space with a DI container.

Mobile objects ... move. So they aren't in a consistent memory space, and so don't have a consistent DI container.

In fact, because of the way we use DI scopes (to support EF, etc.), CSLA domain object graphs move between DI containers quite often during their lifecycle - sometimes across network boundaries too.

In most cases, services that exist on a server won't exist on the client. In fact, the underlying libraries and types won't even exist - things like data access technologies usually aren't available on a wasm client for example.

So if we relied on ctor injection, you'd inject your database connection and similar types that exist only on the server, preventing the domain class from being instantiated at all on the client where those types don't exist at all, and where even if the types _did_ exist you'd never want to register them as services.

If all that isn't enough, DI is much slower to create instances than simple activation of a type. So if we _did_ use DI to create business types we'd all have to take that performance hit.

**Reply by GreatBarrier86** [reply]
In our implementation, our BOs exist in our BusinessObject library that is part of both the client and the server project. Our UI binds to the BusinessBase item that is fetched/created so I’m not sure where the issue is. The server has to have the dependencies to execute the fetch and the client has to have them so the UI can use the BO. 

when would certain libraries not be available?

**Reply by hurcane** [reply]
In our implementation, we use the data portal factory pattern. The data access code is referenced in the server project, but not in the client project. I think that's the scenario Rocky is referring to when he mentions services not being available on the client.

**Reply by GreatBarrier86** [reply]
Ok. Understood. But I don’t think that changes anything if I understand it correctly. Dependencies that are necessary for each side are going to be available in the necessary projects. 

In my scenario, the database connection info type WILL be available client side, per our design. Im not able to visualize a situation where you would have a dependency reference but no available type. 

If I define a type as a dependency, then I obviously have to reference the project that contains the type.

**Reply by StefanOssendorf** [reply]
No. The dependency is not necessarily available. If you have a library which only provides the interfaces and the actual implementation is only available on the server. The client knows the interfaces but no actual class who implements it.

**Reply by StefanOssendorf** [reply]
And just to clarify: in my scenario at work. Not all services are available on both sides. 

Hence Rocky does not support it, as far as I understand, to not make you wonder why you have to make so many unnecessary registrations on both sides.

**Reply by GreatBarrier86** [reply]
So are you saying then that there may be cases where the client may have NO implementation at all?

if that’s the case, then shouldn’t that code be in libraries that aren’t even available client side? Or are we saying it’s acceptable for a Business Object to have a reference to an interface that is not implemented client side (won’t ever be set/injected)

**Reply by StefanOssendorf** [reply]
It is in a library that is not available to the BO. But the BO has to call something to be able to initiate save/update/delete which goes to a database.

**Reply by GreatBarrier86** [reply]
Agreed, but any side that’s calling would obviously have an implementation registered, would it not?

**Reply by StefanOssendorf** [reply]
Yes, obviously it would. Ctor injection is only possible for services who are available on both sides. And that would not be obvious from the start and I think it would do more harm to users learning that the hard way. That's currently using the service locator. I personally don't like the service locator either but it has it's used and I think here is a good place.

**Reply by GreatBarrier86** [reply]
So if the dependencies are currently being injected using the `[Inject]` attribute in the various DP methods, then wouldn't they be equally as available in the ctor?

**Reply by rockfordlhotka** [reply]
Data portal operation methods are _only_ invoked on the logical server, and they are invoked by the data portal. CSLA includes code to do that parameter injection, a concept that's not very common, but works really well in this context.

The original implementation of the `IUseApplicationContext`, which is basically property injection, was naive and when the object graph switched DI contexts, we walked through the entire graph to update all the references to ApplicationContext. That was a performance problem, so now the graph has one reference held by the root object, so we can just change that one reference.

If we allowed some sort of property injection, we'd be back to walking the whole graph on each DI context switch, reflecting against every instance to see if it had any injectable properties. That'd almost certainly be slower than the (comparatively) quick check for `IUseApplicationContext` on each instance.

**Reply by StefanOssendorf** [reply]
No. The ctor demands everything at construction. The methods only when really invoked.
Until the method is invoked the code doesn't care it's not available. Where the ctor would fail immediately even if the service is not needed.

**Reply by rockfordlhotka** [reply]
> Agreed, but any side that’s calling would obviously have an implementation registered, would it not?
>
> So are you saying then that there may be cases where the client may have NO implementation at all?
> 
> if that’s the case, then shouldn’t that code be in libraries that aren’t even available client side? Or are we saying it’s acceptable for a Business Object to have a reference to an interface that is not implemented client side (won’t ever be set/injected)

DI is somewhat "viral". If your domain class requires an `IDal` service, and `IDal` requires a `SqlConnection`, all those services get instantiated in order to inject the `IDal` instance into your domain class.

On a Blazor wasm client, you can't have a `SqlConnection`, and so there'd be no way for DI to resolve `IDal` - unless you created a _different_ "dummy" `IDal` implementation for the client as opposed to the server. That seems like a lot of busywork that everyone would have go through for their services. And for nothing, because it isn't a valid concept in the CSLA architecture for client-side code to interact with a data access layer. That's only valid within the context of the logical server.

Also, flip this around, and suppose someone starts injecting UI elements like pages or viewmodels into their domain types. Those UI types wouldn't exist on a server, and so again, we'd be saying that everyone needs to create dummy implementations of their viewmodels or other UI types on the server so DI could resolve the services. Services that obviously can't be _used_ while on the logical server.

**Reply by GreatBarrier86** [reply]
So that means we just have to make the assumption that any dependency backing fields that were used by the server will simply happen to never be touched by the client? Basically, it's the responsibility of the developer to know that certain fields can't be used by the client?

**Reply by StefanOssendorf** [reply]
IMHO it's bad to even have services in fields. We used that and got a lot of errors. It's not cslas fault but our bad understanding of the client/server side stuff.

**Reply by rockfordlhotka** [reply]
> So that means we just have to make the assumption that any dependency backing fields that were used by the server will simply happen to never be touched by the client? Basically, it's the responsibility of the developer to know that certain fields can't be used by the client?

Parameters injected into data portal operation methods should never be stored as class-level fields or properties. So this should not be a source of confusion.

**Reply by GreatBarrier86** [reply]
The issue we have is that our classes are extremely tightly coupled which is a massive no-no with Dependency Injection, making the entire conversion somewhat awkward and fudgy. Because of that, we have requirements that don't always make a ton of sense from a 3rd party.

**Given that, I guess we're saying this:**
1. Any dependencies needed server-side will be injected through the DP methods.
2. Any dependencies needed client-side will be accessed through ApplicationContext.GetRequiredService()

**Reply by StefanOssendorf** [reply]
Yes. I would say: 
1. Prefer: Request dependencies injected in the DP method
2. Like you wrote
3. Requesting a dependency via application context is regardless of server/client always the last way to get something you need xD

**Comment by GreatBarrier86** [comment]
I think in principlel, it would be this. Basically use it if you already have it somehow and only get if it it's null. 

![image](https://github.com/MarimerLLC/csla/assets/29482393/1d241da1-210a-441e-91d2-d40c591135cd)

**Reply by rockfordlhotka** [reply]
I _think_ that should work. Usually at least. Maybe always.

The big thing is that you don't want to end up holding a reference to a service from a DI scope that's no longer valid. I don't think that should happen as the object graph flows through the data portal (local in particular), though there is a setting for LocalProxy that turns off cloning of the graph, and then I'm _sure_ this would fail.

**Reply by GreatBarrier86** [reply]
We have not changed that parameter for the local proxy, so we should be fine. Thanks for the help everyone.
